pkg/cli/image/mirror/file.go

```diff
@@ -0,0 +1,441 @@
+	"github.com/docker/distribution/manifest"
+	"github.com/docker/distribution/reference"
+	godigest "github.com/opencontainers/go-digest"
+)
+
+type fileDriver struct {
+	BaseDir string
+}
+
+func (d *fileDriver) Repository(ctx context.Context, server *url.URL, repoName string, insecure bool) (distribution.Repository, error) {
+	klog.V(3).Infof("Repository %s %s", server, repoName)
+
+	ref, err := reference.Parse(repoName)
+	if err != nil {
+		return nil, err
+	}
+	named, ok := ref.(reference.Named)
+	if !ok {
+		return nil, fmt.Errorf("%s is not a valid repository name", repoName)
+	}
+
+	repo := &fileRepository{
+		repoName: named,
+		repoPath: repoPathForName(repoName),
+		basePath: d.BaseDir,
+	}
+	return repo, nil
+}
+
+func repoPathForName(repoName string) string {
+	return strings.ReplaceAll(repoName, "/", string(filepath.Separator))
+}
+
+type fileRepository struct {
+	basePath string
+	repoPath string
+	repoName reference.Named
+}
+
+// Named returns the name of the repository.
+func (r *fileRepository) Named() reference.Named {
+	return r.repoName
+}
+
+// Manifests returns a reference to this repository's manifest service.
+// with the supplied options applied.
+func (r *fileRepository) Manifests(ctx context.Context, options ...distribution.ManifestServiceOption) (distribution.ManifestService, error) {
+	return &fileManifestService{r: r}, nil
+}
+
+// Blobs returns a reference to this repository's blob service.
+func (r *fileRepository) Blobs(ctx context.Context) distribution.BlobStore {
+	return &fileBlobStore{r: r}
+}
+
+// Tags returns a reference to this repositories tag service
+func (r *fileRepository) Tags(ctx context.Context) distribution.TagService {
+	return &fileTagStore{r: r}
+}
+
+type fileTagStore struct {
+	r *fileRepository
+}
+
+// Get retrieves the descriptor identified by the tag. Some
+// implementations may differentiate between "trusted" tags and
+// "untrusted" tags. If a tag is "untrusted", the mapping will be returned
+// as an ErrTagUntrusted error, with the target descriptor.
+func (s *fileTagStore) Get(ctx context.Context, tag string) (distribution.Descriptor, error) {
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "manifests", tag)
+	fi, err := os.Lstat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return distribution.Descriptor{}, distribution.ErrBlobUnknown
+		}
+		return distribution.Descriptor{}, err
+	}
+	if fi.Mode()&os.ModeSymlink == 0 {
+		return distribution.Descriptor{}, fmt.Errorf("not a symlink")
+	}
+	target, err := os.Readlink(path)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+	return distribution.Descriptor{
+		Digest: godigest.Digest(filepath.Base(target)),
+		Size:   fi.Size(),
+	}, nil
+}
+
+// Tag associates the tag with the provided descriptor, updating the
+// current association, if needed.
+func (s *fileTagStore) Tag(ctx context.Context, tag string, desc distribution.Descriptor) error {
+	panic("not implemented")
```

Comment:
nit: I would rather see an error, or a little description for why panic is most appropriate in the commit message.